File: src/state.js
const fs = require('fs');
const path = require('path');
const config = require('./config');

const DEFAULT_STATE = {
  google: {
    pageToken: null,
  },
  notion: {
    // pageId/databaseItemId -> ISO last_edited_time 문자열
    lastEditedTimes: {},
  },
};

function load() {
  try {
    const raw = fs.readFileSync(config.stateFilePath, 'utf8');
    const parsed = JSON.parse(raw);
    return {
      google: { ...DEFAULT_STATE.google, ...parsed.google },
      notion: {
        ...DEFAULT_STATE.notion,
        ...parsed.notion,
        lastEditedTimes: { ...parsed.notion?.lastEditedTimes },
      },
    };
  } catch (err) {
    if (err.code !== 'ENOENT') {
      console.warn('[state] 기존 상태 파일을 읽는 중 문제가 발생해 기본값으로 시작합니다:', err.message);
    }
    return JSON.parse(JSON.stringify(DEFAULT_STATE));
  }
}

function save(state) {
  const dir = path.dirname(config.stateFilePath);
  fs.mkdirSync(dir, { recursive: true });
  fs.writeFileSync(config.stateFilePath, JSON.stringify(state, null, 2), 'utf8');
}

module.exports = { load, save };

File: src/state.py
"""폴링 진행 상태(Google pageToken, Notion last_edited_time)를 저장하는 모듈."""
import json
from pathlib import Path

STATE_PATH = Path(__file__).resolve().parent.parent / "data" / "state.json"

DEFAULT_STATE = {
    "google": {"pageToken": None},
    "notion": {"lastEditedTimes": {}},
}


def load() -> dict:
    try:
        parsed = json.loads(STATE_PATH.read_text(encoding="utf-8"))
    except (FileNotFoundError, json.JSONDecodeError):
        return json.loads(json.dumps(DEFAULT_STATE))

    google = {**DEFAULT_STATE["google"], **parsed.get("google", {})}
    notion = {
        **DEFAULT_STATE["notion"],
        **parsed.get("notion", {}),
        "lastEditedTimes": dict(parsed.get("notion", {}).get("lastEditedTimes", {})),
    }
    return {"google": google, "notion": notion}


def save(state: dict) -> None:
    STATE_PATH.parent.mkdir(parents=True, exist_ok=True)
    STATE_PATH.write_text(json.dumps(state, indent=2, ensure_ascii=False), encoding="utf-8")
